WIP: refactor

KuabeM opened the pull request:
To make the discussion easier: is this approach pointing into the right direction? One struct for each type of input which does the parsing.
My idea is then to implement a third struct which implements the actual features like `replace`, `stitch` and so on and also make `From` traits to switch between the representations easily. With the third struct, we only have one implementation of the actual features.

drahnr commented:
You have to define `FromFile` and `ToFile` (if it makes sense, I am not sure it is, just an idea).

drahnr commented:
The <kbd>Update branch</kbd> button is fucked, while no merge commits are allowed, it still does a darn merge...

drahnr commented:
First idea, but I think we should try and do something more rustic with generics and trait bounds, the traits themself might be ok-ish I think.

KuabeM commented:
This brings me to the brink of my rust knowledge... My understanding:
The `ops` trait has to be implemented for each format (binary, intelHex, elf etc.) and does the actual work. The `FromBytes` and `ToBytes` trait is just for conversion between those formats. In the end, we have different implementations for each format right? 
With generics: this would enable having one generic implementation of the Ops for all formats. The trait bound (I'm not quite sure if I really mean the bounds) would be `toBytes` and `fromBytes`?
I'm not sure if I'm gonna be of that much help on that matter :disappointed: 

> The Update branch button is fucked, while no merge commits are allowed, it still does a darn merge...

so always do a `git rebase origin/master` (or whatever branch) and a `git push --force` afterwards?

drahnr commented:
You'll learn faster than you think ;) - generics aren't that much, nor are trait bounds. Just call me from Wednesday on, I am on vacation now ;)

Yep, always do a manual rebase and force push.

There is a missing piece though. Actually I'd lke the intermediate representation (abbrev IR) to be typed, so it should be a `BareMetalBinary` and not just a `BytesMut` (we can think of something else later if there is a need).
If you go from there then things become easier and all ops only need to be only implemented for that, and convert the rest into our IR and go from there.

drahnr commented:
Compiling, yay! :+1:

drahnr commented:
Maybe look into https://crates.io/crates/tree_magic for determining the file type - we only need ASCII/txt and binary, everything else, maybe topped by file extension interpretation.

KuabeM commented:
> Compiling, yay! +1

I :heart: `unimplemented!()` :joy:

KuabeM commented:
What do you think of the `from_content_alt()`? I think it's easier to just pass the path, save us to read the whole file on our own and then pass the first bytes...

drahnr commented:
Took a few minutes, and I realized that the commandline for grafting is not intuitive. The naming and too many options make it difficult.
Idea:
Use a different rust slice like syntax with `<start>..<end>` or `<start>..<including_end>` ranges, I think that could help making it more natural.

Also: grafting doesn't seem to work yet. (manual verification using `hexyl`)

KuabeM commented:
> I'd prefer to not do any file io in tests, but use slices / raw str directly as test input vectors. File io errors are not part of the unit you want to test :) ci should just run the final binary as integration test.
> Any thoughts on that?

Yep makes sense, I already implemented stitch and stance that way. Was just faster yesterday to do the io :)

What do you think of these tests?

KuabeM commented:
> Took a few minutes, and I realized that the commandline for grafting is not intuitive. The naming and too many options make it difficult.
> Idea:
> Use a different rust slice like syntax with `<start>..<end>` or `<start>..<including_end>` ranges, I think that could help making it more natural.

Any idreas on how to achieve that with docopt? can't find anything that allows slices. My idea: use a struct or a string and parse this into the needed start and end.

drahnr commented:
> > Took a few minutes, and I realized that the commandline for grafting is not intuitive. The naming and too many options make it difficult.
> > Idea:
> > Use a different rust slice like syntax with `<start>..<end>` or `<start>..<including_end>` ranges, I think that could help making it more natural.

Made a mistake here, the second one should have been the stabilized inclusive range syntax `a..=b`

> Any idreas on how to achieve that with docopt? can't find anything that allows slices. My idea: use a struct or a string and parse this into the needed start and end.

Yep, exactly that :) - combined with serde for decoding it straight from a string, which is used implicitly by docopt.rs.

KuabeM commented:
I introduced a `Range` struct holding `start` and `end` as `ByteOffsets` such that we can use things like `13Ki..5G` directly as input.

Right now it only takes ranges as `<start>..<end>`, we therefore 'lost' the feature of having size or end specified, but this also means that there are less combinations of flags. Maybe we can find a way to include a pattern for `start` and `size` in another way.

What do you think?

KuabeM commented:
I like the `start+size` idea and since I couldn't think of anything else, I implemented it. We now accept either `<start>..<end>` or `<start>+<size>` with all possible units. I cleaned up the main and docopt stuff and it's now a lot cleaner.

Range has now start and size as `stance` and `graft` use those as inputs. The input `end` has to be converted to size therefore which happens directly in the Deserialize together with the check for `start < end`.

Right now, we don't have a lot of tests, we basically only cover the main functions of `AnnotatedBytes` and the `intelhex` module (where we also have io stuff..). Do you think we should add more?

Another thing came to my mind: how about ranges as hex numbers, e.g. 0x10..0x40? Worth implementing?

drahnr commented:
0x prefix as hex handling sounds like good feature. I hate having to pull the calculator to convert them, so +1, maybe even as part of `ByteOffset` deserialization?

drahnr commented on src/range.rs:
```diff
@@ -37,7 +37,7 @@ impl<'de> de::Deserialize<'de> for Range {
                 lazy_static! {
                     static ref REGEX: Regex = Regex::new(
-                        r"^(([0-9]+)((?:[KMGTE]i?)?))(\.\.|\+)(([0-9]+)((?:[KMGTE]i?)?))$"
+                        r"^((0x)?([0-9]+)((?:[KMGTE]i?)?))(\.\.|\+)((0x)?([0-9]+)((?:[KMGTE]i?)?))$"
```
What about `0xab` or `0xAB`?

KuabeM replied:
stupid me..

drahnr commented on Cargo.toml:
```diff
@@ -20,14 +21,16 @@ log = "0.4"
 env_logger = "0.5"
 lazy_static = "1.1.0"
 bytes = "0.4.9"
-ring = "0.12.1"
```
Is untrusted still necessary? That was only for ring and according to it's author none of the inputs for the common api requires untrusted no more.

drahnr commented on README.md:
```diff
@@ -5,51 +5,43 @@
 
 * [x] cut off a binary at specific start and end/size
-* [x] Add signature verification and appendix features (using preferably [ring] and [webpki])
+* [ ] Add signature verification and appendix features (using preferably [ring] and [webpki])
```
Maybe just delete this line entirely

drahnr commented on src/byte_offset.rs:
```diff
@@ -94,27 +157,33 @@ impl<'de> de::Deserialize<'de> for ByteOffset {
                 lazy_static! {
-                    static ref REGEX: Regex = Regex::new(r"^([0-9]+)((?:[KMGTE]i?)?)$").unwrap();
+                    static ref REGEX: Regex =
+                        Regex::new(r"^(0x)?([0-9]+)((?:[KMGTE]i?)?)$").unwrap();
```
Hex and non unit doesn't make much sense, does it?

KuabeM commented:
I hope I figured out all of the 'complicated' hex parts.. 

I added some simple tests for `FromStr` of `ByteOffset`. I hat a biref look at [serde-test](https://docs.rs/serde_test/1.0.90/serde_test/) but can't figure it out now.. you think that'd be useful?

drahnr commented:
Not sure if `serde-test` is what we need. It creates an IR repr as far as I understand for which you can unit test, but we don't really need testing to that extend, it should be enough to test all possible combinations which we expect our regex to match / error?

drahnr commented:
Tried to `stitch` some file and it did not work, it always fails with
```sh
# cargo run -- stitch --input=./grafted.bin --offset=0 --input=./stanced.bin --offset=3351 --output=x5.bin
Error: Offset 0 smaller than current file 350
```

It might make sense to run the binaries as part of the ci with all flags specified.

KuabeM commented:
I just had a quick look at it, the problem lies in the `main.rs`. All offsets are zero, seems that they don't get parsed correctly or the zipping is not working. Will have a look tonight :)

KuabeM commented:
Found the error, the regex was not the right one. I simplified `Deserialize` by using the `FromStr` there, so we only have the regex stuff once. I also added a test for ByteOffset.

drahnr commented on src/main.rs:
```diff
@@ -302,4 +302,45 @@ mod test {
+                "scalpel",
+                "stitch",
+                "--input",
+                "byets",
```
typo? doesn't matter, just cought my attention

drahnr commented on src/byte_offset.rs:
```diff
@@ -156,53 +156,7 @@ impl<'de> de::Deserialize<'de> for ByteOffset {
-                    })
-                    .map_err(|e| E::custom(format!("{:?}", e)))?;
-                Ok(byte_offset)
+                ByteOffset::from_str(value).map_err(|e| E::custom(format!("{:?}", e)))
```
:+1:

drahnr commented:
I created a bit of a mess, I force pushed all changes to master and unfortunately I acceidentally released the wrong branch as 1.0.0 (but yanked it already), so this is going to be 2.0.0, sorry.

Generally, I think it might be good to have cargo release be part of ci, maybe a separate release pipeline?
